### ISSUE TITLE
Analytics(session tracking): Add option not to trigger new session on visibilityChange

### ISSUE THREAD

**Issue by germain-receeve** [aws-amplify/amplify-js#4517]
**Is your feature request related to a problem? Please describe.**
Yes, in fact we want to track only when the user entered our page and closed it. At the moment, `_session.start` and ``_session.close` event are also emitted when switching tab for example because there is a watcher on page visibility (hidden or not)

**Describe the solution you'd like**
An option for the `autotrack` feature that would allow us to disable this behaviour.
Could be for example `trackVisibilityChange` which would 

**Describe alternatives you've considered**
Tracking the session event ourselves. Sadly, it's not feasible because the `immediate` event firing is only available internally in the lib, therefore it's not possible to send an event during `beforeUnload` event when user leaves the page.

**Additional context**
Line responsible for tracking `visibilityChange` events
https://github.com/aws-amplify/amplify-js/blob/8d2ba6e85031a7880d2b573e1f68108d22a7de54/packages/analytics/src/trackers/SessionTracker.ts#L166

**Comment by ericclemmons** [comment]
You're right, the `_session.stop` event is called when visibility is lost.

https://github.com/aws-amplify/amplify-js/blob/8d2ba6e85031a7880d2b573e1f68108d22a7de54/packages/analytics/src/trackers/SessionTracker.ts#L81-L87

Is the problem that you can't differentiate the visibility events vs. close events? (For example, `_session.pause` and `_session.resume` vs. `_session.start` and `_session.stop`)

Or is the problem that it's creating noise and you'd like to filter these out entirely?

**Comment by germain-receeve** [comment]
I would say it's more that it's making noise since in our case we are mainly interested by the user starting to use the site and effectively quitting it.
If we could differentiate those events like you suggested, that would work for us as well 👍

**Comment by ericclemmons** [comment]
Great feedback @germain-receeve!

I'd like to proceed with getting some internal feedback with differentiating those events vs. filtering, and can even put together a PR for you.

Here's my rationale:

- Adding more API options would require more consideration & a higher bar for testing & support.
- Differentiating these events, in my opinion, has a lot of value in data-slicing: you'll be able to understand why there was a lull in events vs. seeing nothing for a long period of time.

**Comment by ericclemmons** [comment]
@germain-receeve Alright, I quickly put together https://github.com/aws-amplify/amplify-js/pull/4525 pretty easily, so we can have an internal discussion to ensure nothing blows up 🤞 

Otherwise, we'll invest some time to look into the filtering option you suggested.

Thanks again for bringing this to our attention!

**Comment by ericclemmons** [comment]
@germain-receeve I just wanted to let you know that we've been having several discussions behind the scenes:

- There's validity in introducing new events, but would need to be considered for all platforms (including iOS & Android).  **Would this still be a "noise" problem for you**, or is it not considered noise since they would be different from start/stop?

- Introducing new events would likely affect customer reporting, which could warrant a major version bump. This would affect the timeline.

- A shorter-term solution would be to introduce a means to filter events on your end. That approach would be a separate path to review the potential API.

	> An option for the autotrack feature that would allow us to disable this behaviour.
Could be for example trackVisibilityChange option which would not add the event listener if set to false.

**Comment by germain-receeve** [comment]
@ericclemmons Thanks for the feedback about the current state of the issue.

- No, they would not be noise anymore since we would only take into account the `start/stop` events and ignore the `pause/resume` that are not relevant to our use case.

- Sadly you are right, this will probably require a major version and therefore take some time.

- For now I introduced a hack on our end that listen to the `visibilityChange` event from the browser and stop its propagation. Therefore amplify don't send session start/stop when switching tabs. 
Having a flag that make the listener in the lib optional would be really nice. We could get rid of the hack :)

**Comment by ericclemmons** [comment]
@iartemiev Can you weigh in on the option for introducing a flag vs. customers intervening with `event.stopPropagation()`?

**Comment by damien-monni** [comment]
> @ericclemmons
> 
> * For now I introduced a hack on our end that listen to the `visibilityChange` event from the browser and stop its propagation. Therefore amplify don't send session start/stop when switching tabs.

Hi! Can you share how did you override the visibilityChange event? I am trying to prevent sessions stops on tab change.

**Comment by germain-receeve** [comment]
@damien-monni 
Sure, here is my solution to circumvent the problem 😄 

```
<script lang="ts">
import { Component, Vue } from "vue-property-decorator";

@Component
export default class App extends Vue {
  private visibilityChangeName = this.getVisibilityChangeName();

  created() {
    /**
     ** Prevent pinpoint from sending session start/stop when tab switch to foreground/background
     ** https://github.com/aws-amplify/amplify-js/blob/master/packages/analytics/src/trackers/SessionTracker.ts
     **/
    if (this.visibilityChangeName) {
      window.addEventListener(
        this.visibilityChangeName,
        this.stopEventPropagation,
        true
      );
    }
  }

  destroy() {
    if (this.visibilityChangeName) {
      window.removeEventListener(
        this.visibilityChangeName,
        this.stopEventPropagation,
        true
      );
    }
  }

  // Return event change listener name (https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API)
  getVisibilityChangeName(): string | undefined {
    if (typeof document.hidden !== "undefined") {
      return "visibilitychange";
      //@ts-ignore
    } else if (typeof document.msHidden !== "undefined") {
      return "msvisibilitychange";
      //@ts-ignore
    } else if (typeof document.webkitHidden !== "undefined") {
      return "webkitvisibilitychange";
    }
  }

  stopEventPropagation(e: Event): void {
    e.stopPropagation();
  }
}
</script>
```

**Comment by damien-monni** [comment]
@germain-receeve I needed to pass true to the third addEventListener param. It works perfectly now, thank you!
